Amplify/Categories/DataStore/Model/Internal/ModelProvider.swift

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// Protocol used as a marker to detect when the type is a `LazyReference`.
+/// Used to retrieve either the `reference` or the `identifiers` of the Model directly, without having load a not
+/// loaded LazyReference. This is useful when translating the model object over to the payload required for the
+/// underlying storage, such as storing the values in DataStore's local database or AppSync GraphQL request payload.
+///
+///
+/// - Warning: Although this has `public` access, it is intended for internal & codegen use and should not be used
+/// directly by host applications. The behavior of this may change without warning. Though it is not used by host
+/// application making any change to these `public` types should be backward compatible, otherwise it will be a breaking
+/// change.
+public protocol LazyReferenceMarker {
+    associatedtype ModelType: Model
+    
+    var reference: ModelType? { get }
+    
+    var identifiers: [LazyReferenceIdentifier]? { get }
+}
+
+/// State of the ModelProvider
+///
+/// - Warning: Although this has `public` access, it is intended for internal & codegen use and should not be used
+/// directly by host applications. The behavior of this may change without warning. Though it is not used by host
+/// application making any change to these `public` types should be backward compatible, otherwise it will be a breaking
+/// change.
+public enum ModelProviderState<Element: Model> {
+    case notLoaded(identifiers: [LazyReferenceIdentifier]?)
+    case loaded(Element?)
+}
+
+/// - Warning: Although this has `public` access, it is intended for internal & codegen use and should not be used
+/// directly by host applications. The behavior of this may change without warning. Though it is not used by host
+/// application making any change to these `public` types should be backward compatible, otherwise it will be a breaking
+/// change.
+public protocol ModelProvider {
+    associatedtype Element: Model
+    
+    func load() async throws -> Element?
+    
+    func getState() -> ModelProviderState<Element>
+}
+
+/// - Warning: Although this has `public` access, it is intended for internal & codegen use and should not be used
+/// directly by host applications. The behavior of this may change without warning. Though it is not used by host
+/// application making any change to these `public` types should be backward compatible, otherwise it will be a breaking
+/// change.
+public struct AnyModelProvider<Element: Model>: ModelProvider {
+    
+    private let loadAsync: () async throws -> Element?
+    private let getStateClosure: () -> ModelProviderState<Element>
```

Comment:
this could be private `let provider: ModelProvider<Element>` instead of copying each function you need to delegate. Then in `load()` you call `provider.load()` and so on

Reply:
here are a few attempts
<img width="700" alt="image" src="https://user-images.githubusercontent.com/1365977/204378688-56922302-8dc5-4575-bb32-be96b157b2e6.png">

<img width="852" alt="image" src="https://user-images.githubusercontent.com/1365977/204379162-6a6d0522-ed46-428a-a63c-6af305b8c2bf.png">